pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,346 @@
+		return "", "", invalidRepoTagErr
+	}
+}
+
+// flushMap deletes all they keys from a map
+func flushMap(repoList map[string]string) {
+	for key := range repoList {
+		delete(repoList, key)
+	}
+}
+
+// findRepoDeps accepts an image repository name and a list of build
+// configurations and returns the dependency tree of the specified
+// image repository
+func findRepoDeps(name, namespace, tag string, all bool, buildConfigList []buildapi.BuildConfig) *ImageRepo {
+	root := &ImageRepo{
+		Name:      name,
+		Namespace: namespace,
+		Tags:      []string{tag},
+	}
+
+	// Search all build configurations in order to find the image
+	// repositories depending on the specified image repository
+	var outputName, outputTag string
+	for _, cfg := range buildConfigList {
+		for _, tr := range cfg.Triggers {
+			// If an ImageChangeTrigger has been specified (ie. it's not nil), use its From
+			// and Tag fields to compare against the specified image repository
+			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all || tag == tr.ImageChange.Tag) {
+				// Either To & Tag or DockerImageReference will be used as output
+				if cfg.Parameters.Output.To.Name != "" {
+					outputName = cfg.Parameters.Output.To.Name
+					outputTag = cfg.Parameters.Output.Tag
+					if cfg.Parameters.Output.To.Namespace != "" {
+						namespace = cfg.Parameters.Output.To.Namespace
+					}
+				} else {
+					ref, err := imageapi.ParseDockerImageReference(cfg.Parameters.Output.DockerImageReference)
+					checkErr(err)
+					outputName = ref.Name
+					outputTag = ref.Tag
+				}
+
+				// Build all children and their dependency trees recursively
+				child := findRepoDeps(outputName, namespace, outputTag, all, buildConfigList)
+
+				// Add edge between root and child
+				root.Edges = append(root.Edges, NewEdge(cfg.Name, child.Name))
+
+				// If the user specifies the --all flag and the child depends
+				// on root via more than one tag, multiple instances of the child
+				// will make it into root.Children
+				if all {
+					cont := false
+					for _, repo := range root.Children {
+						if repo.Name == child.Name {
+							// Just pass the tag along and discard the current child
+							repo.Tags = append(repo.Tags, child.Tags...)
+							cont = true
+							break
+						}
+					}
+					if cont {
+						// Do not append this child in root.Children. It's already in there
+						continue
+					}
+				}
+
+				root.Children = append(root.Children, child)
+			}
+		}
+	}
+
+	return root
+}
+
+// dotDump dumps the given image repository tree in DOT syntax
+func dotDump(root *ImageRepo, g *dot.Graph, graphName string) string {
+	if root == nil {
+		return ""
+	}
+
+	// Add current node
+	attrs := make(map[string]string)
+	for i, tag := range root.Tags {
+		// Values in DOT maps need to be quoted
+		attrs["tag"+strconv.Itoa(i)] = strconv.Quote(tag)
+	}
+	attrs["namespace"] = strconv.Quote(root.Namespace)
+	rootName := strings.Replace(root.Name, "-", "_", -1)
```

Comment:
is the hyphen("-") the only character that can be invalid in a dot node? If not, it may be good to abstract this into a function that turns a K8s name into a valid dot name.